### ISSUE TITLE
Memory usage grows over time and ends with memory error

### ISSUE THREAD

**Issue by berniejerom** [iperov/DeepFaceLab#116]
## Expected behavior
memory usage probaly shouldn't rise a few GB over time.

## Actual behavior
I am able to run a few hundreds epochs but then I get the memory error.
Training window process memory usage is stable, but the other python process can rise up to 2GB each.
 
![bern](https://user-images.githubusercontent.com/46869905/51443220-de0d1a80-1ce5-11e9-83b7-fa83e655c2cd.PNG)

I get memory error when the app uses about ~8.5GB, even though some of RAM is still available.
GPU mem usage looks to be stable - about 10GB

```
 g [#004905][1596ms] src_loss:0.842 dst_loss:0.318
Traceback (most recent call last):

  File "D:\deepFaceLab\mainscripts\Trainer.py", line 71, in trainerThread
    loss_string = model.train_one_epoch()

  File "D:\deepFaceLab\models\ModelBase.py", line 308, in train_one_epoch
    self.last_sample = self.generate_next_sample()

  File "D:\deepFaceLab\models\ModelBase.py", line 301, in generate_next_sample
    return [next(generator) for generator in self.generator_list]

  File "D:\deepFaceLab\models\ModelBase.py", line 301, in <listcomp>
    return [next(generator) for generator in self.generator_list]

  File "D:\deepFaceLab\samples\SampleGeneratorFace.py", line 55, in __next__
    return next(generator)

  File "D:\deepFaceLab\utils\iter_utils.py", line 57, in __next__
    gen_data = self.cs_queue.get()

  File "c:\users\bern\appdata\local\programs\python\python36\Lib\multiprocessing\queues.py", line 94, in get
    res = self._recv_bytes()
  File "c:\users\bern\appdata\local\programs\python\python36\Lib\multiprocessing\connection.py", line 216, in recv_bytes
    buf = self._recv_bytes(maxlength)
  File "c:\users\bern\appdata\local\programs\python\python36\Lib\multiprocessing\connection.py", line 318, in _recv_bytes
    return self._get_more_data(ov, maxsize)
  File "c:\users\bern\appdata\local\programs\python\python36\Lib\multiprocessing\connection.py", line 344, in _get_more_data
    f.write(ov.getbuffer())
MemoryError
Done.
```

## Steps to reproduce
Train any model, I think it also occurs for liaef and DF, sometimes it just takes more time.

## Other relevant information
Python 3.6.5 64bit
Windows 10
1080TI
16 GB + Swap 10GB
CUDA 9.0
CUDNN v7.4.1.5 also tried newest one
Note: I've changed save period to 2min.

**Comment by TooMuchFun** [comment]
Same for me on Ubuntu w/ 16GB and a 1070. It will last ~700 iterations or 10 minutes before the TF session dies. The python process remains on, tqdm updating the status line, but the preview no longer responds and the system becomes very unstable.

**Comment by iperov** [comment]
not you

**Comment by iperov** [comment]
TF session dies - another issue.

**Comment by berniejerom** [comment]
src 319 files, 30MB
dst 1570 files, 122MB
all 4 python processes are growing except the one which is training preview

with CPU I think it starts about ~7GB and it works though it takes so long that I cannot test if there is some memory leak

**Comment by iperov** [comment]
these 4 processes generate samples for training. 2 for src and 2 for dst.

I changed ModelBase.py to this code
```
    def train_one_epoch(self):    
...
        while True: #<- new line
            self.last_sample = self.generate_next_sample() 
```
to generate samples as fast as possible in loop before training start
but no memoryleak 
![2019-01-21_00-30-05](https://user-images.githubusercontent.com/8076202/51444645-b03eca80-1d13-11e9-8e00-62ba9dabb910.gif)
cannot reproduce in my prebuilt windows binary.

**Comment by berniejerom** [comment]
did the same thing and it is growing
![obraz](https://user-images.githubusercontent.com/46869905/51444746-dc4e5180-1cfb-11e9-8d8b-15ea8407de54.png)
![obraz](https://user-images.githubusercontent.com/46869905/51444752-f6882f80-1cfb-11e9-955e-378c1ac2a1fb.png)

I will try to narrow the scope in a next few days, thanks for the support

**Comment by iperov** [comment]
and why are you not using prebuilt windows binary?

**Comment by berniejerom** [comment]
Sometimes I need to customize existing code like with the saving interval, so it is nice to use repository instead, but actually the need was caused by the out of memory problem. 

I have no memory leak using prebuild binary, also when I've just copied the code from prebuild _internal_ to deepFaceLab the memory leak was visible so it seems it is related to some external module

**Comment by berniejerom** [comment]
Hi, I have used yolk to compare installed site packages in prebuilt package and downloaded project.
After some trial and error it seems that numpy 1.16 is causing the issue - prebuild package has 1.15.4 version

**Comment by iperov** [comment]
@berniejerom wow interesting.

**Comment by iperov** [comment]
I have upgraded numpy to 1.16 and got memory leak too

**Comment by iperov** [comment]
looks like memory leak caused by np array interprocess pickling

**Comment by iperov** [comment]
https://github.com/numpy/numpy/issues/12822

**Comment by iperov** [comment]
https://github.com/numpy/numpy/issues/12793

they will fix it in 1.16.1
